### PR TITLE
Refactored Interchain Exchange tutorial

### PR DESCRIPTION
**Pull request by fadeev**
Currently refactored tutorial in its entirety is in a single file ([`00-tutorial.md`](https://github.com/cosmos/sdk-tutorials/blob/3019c977f7d04298f83575d159f3da672d04395f/interchain-exchange/tutorial/00-tutorial.md)) to make editing easier. Before merging, it should be split into multiple files.

The idea behind the refactor is to implement each high level (CLI) command first, then implement a function that this command depends on, instead of having multiple sections dedicated to implementing prerequisites. This will make it much easier to record a screencast.

This tutorial still requires text commentary on what's happening.

### CODE REVIEW

File: interchain-exchange/tutorial/00-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,2585 @@
...
+```
+
+You can also use Starport v0.16.2 in a [browser-based IDE](http://gitpod.io/#https://github.com/tendermint/starport/tree/v0.16.2), but this tutorial assumes you are using a local Starport installation. For more installation options, see [install Starport](https://docs.starport.network/intro/install.html).
+
+## Create the Blockchain
+
+Scaffold a new blockchain called `interchange`
+
+```bash
+starport app github.com/username/interchange
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:114]: starport scaffold chain github.com/username/interchange --no-default-module

File: interchain-exchange/tutorial/02-app-init.md [hunk 2]
```diff
@@ -45,12 +45,11 @@ To scaffold two types with create, read, update and delete (CRUD) actions use th
...
 
 ```bash
-starport type sellOrderBook orderIDTrack:int amountDenom priceDenom --indexed --no-message --module ibcdex
-starport type buyOrderBook orderIDTrack:int amountDenom priceDenom --indexed --no-message --module ibcdex
+starport type sellOrderBook amountDenom priceDenom --indexed --no-message --module ibcdex
+starport type buyOrderBook amountDenom priceDenom --indexed --no-message --module ibcdex
 ```
 
 The values are: 
-- `orderIDTrack` an internal counter in the order book to assign the orders an ID.
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/02-app-init.md:53]: `orderIdTrack` an internal counter in the order book to assign the orders an ID

File: interchain-exchange/tutorial/00-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,2585 @@
...
+
+The optional `--ack` flag defines field names and types of the acknowledgment returned after the packet has been received by the target chain. Value of `--ack` is a comma-separated (no spaces) list of names with optional types appended after a colon.
+
+## Cancel messages
+
+Cancelling orders is done locally in the network, there is no packet to send.
+Use the `message` command to create a message to cancel a sell or buy order.
+
+```go
+starport message cancelSellOrder port channel amountDenom priceDenom orderID:int --desc "Cancel a sell order" --module ibcdex
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:171]: starport message cancel-sell-order port channel amountDenom priceDenom orderID:int --desc "Cancel a sell order" --module ibcdex

File: interchain-exchange/tutorial/00-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,2585 @@
...
+starport message cancelBuyOrder port channel amountDenom priceDenom orderID:int --desc "Cancel a buy order" --module ibcdex
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:172]: starport message cancel-buy-order port channel amountDenom priceDenom orderID:int --desc "Cancel a buy order" --module ibcdex

File: interchain-exchange/tutorial/03-orderbook.md [hunk 5]
```diff
@@ -121,36 +162,40 @@ func AppendOrder(book OrderBook, creator string, amount int32, price int32) (Ord
...
+	} else {
+		book.Orders = append(book.Orders, &order)
+	}
 }
 ```
 
 ## Add The Sellorder
 
-Modify the `sellOrderBook.proto` file to have the fields for creating a sell order on the order book.
+Modify the `sellOrderBook.proto` file to add the order book into the sell order book.
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/03-orderbook.md:198]: Modify the `sell_order_book.proto` file to add the order book into the sell order book.

File: interchain-exchange/tutorial/00-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,2585 @@
...
+		PriceDenom: PriceDenom,
+		Book: &book,
+	}
+}
+```
+
+Modify the `sellOrderBook.proto` file to add the order book into the buy order book. The proto definition for the `SellOrderBook` should look like follows:
+
+```proto
+// proto/ibcdex/sellOrderBook.proto
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:769]: // proto/ibcdex/sell_order_book.proto

File: interchain-exchange/tutorial/03-orderbook.md [hunk 7]
```diff
@@ -165,188 +210,107 @@ import "ibcdex/order.proto"; // <--
 message SellOrderBook {
   string creator = 1;
   string index = 2;
-  int32 orderIDTrack = 3;
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/03-orderbook.md:168]: int32 orderIdTrack = 3;

File: interchain-exchange/tutorial/00-tutorial.md [hunk 8]
```diff
@@ -0,0 +1,2585 @@
...
+	book := NewOrderBook()
+	return BuyOrderBook{
+		AmountDenom: AmountDenom,
+		PriceDenom: PriceDenom,
+		Book: &book,
+	}
+}
+```
+
+Modify the `buyOrderBook.proto` file to add an order book to the buy order book.
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:679]: Modify the `buy_order_book.proto` file to have the fields for creating a buy order on the order book.

File: interchain-exchange/tutorial/00-tutorial.md [hunk 9]
```diff
@@ -0,0 +1,2585 @@
...
+
+```proto
+// proto/ibcdex/buyOrderBook.proto
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:682]: // proto/ibcdex/buy_order_book.proto

File: interchain-exchange/tutorial/00-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,2585 @@
...
+  * Transmit an IBC acknowledgement back to the source chain
+* Upon receiving the acknowledgement on the source chain
+  * Create a new order book for sell orders
+
+## `SendCreatePair` Message Handling
+
+`SendCreatePair` function was created during the IBC packet scaffolding. Currently, it creates an IBC packet, populates it with source and target denoms and transmits this packet over IBC. Add the logic to check for an existing order book for a particular pair of denoms.
+
+```go
+// x/ibcdex/keeper/msg_server_createPair.go
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:611]: // x/ibcdex/keeper/msg_server_create_pair.go

File: interchain-exchange/tutorial/00-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,2585 @@
...
+4. Timeout of a packet (on the source chain)
+
+In the following section you'll be implementing packet reception logic in the `OnRecvCreatePairPacket` function and packet acknowledgement logic in the `OnAcknowledgementCreatePairPacket` function. Timeout function will be left empty.
+
+## `OnRecv`
+
+On the target chain when an IBC packet is recieved, the module should check whether a book already exists, if not, create a new buy order book for specified denoms.
+
+```go
+// x/ibcdex/keeper/createPair.go
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:645]: // x/ibcdex/keeper/create_pair.go

File: interchain-exchange/tutorial/00-tutorial.md [hunk 12]
```diff
@@ -0,0 +1,2585 @@
...
+  int32 price = 4;
+}
+```
+
+## `OnAcknowledgement`
+
+On the source chain when an IBC acknowledgement is recieved, the module should check whether a book already exists, if not, create a new sell order book for specified denoms.
+
+```go
+// x/ibcdex/keeper/createPair.go
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:727]: // x/ibcdex/keeper/create_pair.go

File: interchain-exchange/tutorial/00-tutorial.md [hunk 13]
```diff
@@ -0,0 +1,2585 @@
...
+
+* Check that an order book for specified denom pair exists
+* Safely burn or lock tokens
+  * If the token is an IBC token, burn the tokens
+  * If the token is a native token, lock the tokens
+* Save the voucher received on the target chain to later resolve a denom
+* Transmit an IBC packet to the target chain
+
+```go
+// x/ibcdex/keeper/msg_server_sellOrder.go
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:809]: // x/ibcdex/keeper/msg_server_sell_order.go

File: interchain-exchange/tutorial/00-tutorial.md [hunk 14]
```diff
@@ -0,0 +1,2585 @@
...
+## `OnRecv`
+
+When a "sell order" packet is received on the target chain, the module should  ????
+
+- Update the sell order book
+- Distribute sold token to the buyer
+- Send to chain A the sell order after the fill attempt
+
+```go
+// x/ibcdex/keeper/sellOrder.go
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:1035]: // x/ibcdex/keeper/sell_order.go

File: interchain-exchange/tutorial/00-tutorial.md [hunk 15]
```diff
@@ -0,0 +1,2585 @@
...
+	return nil
+}
+```
+
+## `OnAcknowledgement`
+
+Once an IBC packet is processed on the target chain, an acknowledgement is returned to the source chain and processed in `OnAcknowledgementSellOrderPacket`. The module on the source chain will store the remaining sell order in the sell order book and will distribute sold tokens to the buyers and will distribute to the seller the price of the amount sold. On error the module mints the burned tokens.
+
+```go
+// x/ibcdex/keeper/sellOrder.go
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:1249]: // x/ibcdex/keeper/sell_order.go

File: interchain-exchange/tutorial/00-tutorial.md [hunk 16]
```diff
@@ -0,0 +1,2585 @@
...
+	}
+}
+```
+
+## `OnTimeout`
+
+If a timeout occurs, we mint back the native token.
+
+```go
+// x/ibcdex/keeper/sellOrder.go
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:1445]: // x/ibcdex/keeper/sell_order.go

File: interchain-exchange/tutorial/00-tutorial.md [hunk 17]
```diff
@@ -0,0 +1,2585 @@
...
+  return b.Book.appendOrder(creator, amount, price, Increasing)
+}
+```
+
+## `OnTimeout`
+
+If a timeout occurs, we mint back the native token.
+
+```go
+// x/ibcdex/keeper/sellOrder.go
```
Comment (alijnmerchant21) [inline interchain-exchange/tutorial/00-tutorial.md:1741]: // x/ibcdex/keeper/sell_order.go

**Comment by fadeev** [comment]
@alijnmerchant21 thanks for the changes! Do you mind using the suggestion feature next time, which makes it easier to apply changes?

File: interchain-exchange/tutorial/02-app-init.md (hunk 2 above)
Comment (fadeev) [reply interchain-exchange/tutorial/02-app-init.md:53]: This is a comment on a previous version of the tutorial.

File: interchain-exchange/tutorial/03-orderbook.md (hunk 5 above)
Comment (fadeev) [reply interchain-exchange/tutorial/03-orderbook.md:198]: Comment on a previous version of the tutorial.

**Comment by lubtd** [comment]
@alijnmerchant21 thanks for the suggestion. It seems some of them are related to the fact you used Starport on `develop` branch for the tutorial (for the filename changes and command with kebab case for example). FYI we always keep the tutorials adapted to the latest release version on Starport, this is why we still have the old files and commands naming.
For more information: https://github.com/cosmos/sdk-tutorials/pull/744#issuecomment-871156519

**Pull request closed without merging**